src/tools/pkcs11-tool.c

```diff
@@ -5262,11 +5269,96 @@ static int encrypt_decrypt(CK_SESSION_HANDLE session,
 		EVP_PKEY_free(pkey);
 		return 0;
 	}
-	encrypted_len = EVP_PKEY_encrypt_old(encrypted, orig_data, sizeof(orig_data), pkey);
-	EVP_PKEY_free(pkey);
-	if (((int) encrypted_len) <= 0) {
-		printf("Encryption failed, returning\n");
-		return 0;
+	if (mech_type == CKM_RSA_PKCS_OAEP) {
+		EVP_PKEY_CTX *ctx;
+		ctx = EVP_PKEY_CTX_new(pkey, NULL);
+		if (!ctx) {
+			EVP_PKEY_free(pkey);
+			printf("EVP_PKEY_CTX_new failed, returning\n");
+			return 0;
+		}
+		if (EVP_PKEY_encrypt_init(ctx) <= 0) {
+			EVP_PKEY_CTX_free(ctx);
+			EVP_PKEY_free(pkey);
+			printf("EVP_PKEY_encrypt_init failed, returning\n");
+			return 0;
+		}
+
+		if (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_OAEP_PADDING) <= 0) {
+			EVP_PKEY_CTX_free(ctx);
+			EVP_PKEY_free(pkey);
+			printf("set OAEP padding failed, returning\n");
+			return 0;
+		}
+		size_t outlen = sizeof(encrypted);
+		if (EVP_PKEY_encrypt(ctx, encrypted, &outlen, orig_data, sizeof(orig_data)) <= 0) {
+			EVP_PKEY_CTX_free(ctx);
+			EVP_PKEY_free(pkey);
+			printf("Encryption failed, returning\n");
+			return 0;
+		}
+		EVP_PKEY_CTX_free(ctx);
+		EVP_PKEY_free(pkey);
+		encrypted_len = outlen;
+
+	} else {
```

Comment:
I think this should also explicitly check for `CKM_RSA_PKCS` as the decryption does so we can fail earlier if the user provided unsupported mechanism.

Reply:
You are right, we need to check explicitly for CKM_RSA_PKCS and properly handle the case of CKM_RSA_X_509.
I will try to fix this issue during the week.